symphony-bdk-core/src/main/java/com/symphony/bdk/core/util/RetryWithRecovery.java

```diff
@@ -0,0 +1,119 @@
+package com.symphony.bdk.core.util;
+
+import com.symphony.bdk.core.api.invoker.ApiException;
+import com.symphony.bdk.core.config.model.BdkRetryConfig;
+
+import io.github.resilience4j.retry.Retry;
+import io.github.resilience4j.retry.RetryConfig;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Map;
+import java.util.function.Predicate;
+
+/**
+ * This class aims to implement a retry mechanism (on top of a{@link Retry})
+ * with different recovery strategies based on predicates.
+ * @param <T> the type of the object returned by {@link #execute()}
+ */
+@Slf4j
+public class RetryWithRecovery<T> {
```

Comment:
I think that we should try to separate contract from implementation with this feature. Do we want want to make it available for developers? Are we sure that we will keep `resilience4j` as retry library in the future?